handler/src/main/java/io/netty/handler/ssl/OpenSslX509KeyManagerFactory.java

```diff
@@ -0,0 +1,220 @@
+                throws KeyStoreException, NoSuchAlgorithmException, UnrecoverableKeyException {
+            if (providerFactory != null) {
+                throw new KeyStoreException("Already initialized");
+            }
+            if (!keyStore.aliases().hasMoreElements()) {
+                throw new KeyStoreException("No aliases found");
+            }
+
+            kmf.init(keyStore, chars);
+            providerFactory = new ProviderFactory(
+                    ReferenceCountedOpenSslContext.chooseX509KeyManager(kmf.getKeyManagers()),
+                    password(chars), Collections.list(keyStore.aliases()));
+        }
+
+        private static String password(char[] password) {
+            if (password == null || password.length == 0) {
+                return null;
+            }
+            return new String(password);
+        }
+
+        @Override
+        protected void engineInit(ManagerFactoryParameters managerFactoryParameters)
+                throws InvalidAlgorithmParameterException {
+            throw new InvalidAlgorithmParameterException("Not supported");
+        }
+
+        @Override
+        protected KeyManager[] engineGetKeyManagers() {
+            ProviderFactory providerFactory = this.providerFactory;
+            if (providerFactory == null) {
+                throw new IllegalStateException("engineInit(...) not called yet");
+            }
+            return new KeyManager[] { providerFactory.keyManager };
+        }
+
+        OpenSslKeyMaterialProvider newProvider() {
+            ProviderFactory providerFactory = this.providerFactory;
+            if (providerFactory == null) {
+                throw new IllegalStateException("engineInit(...) not called yet");
+            }
+            return providerFactory.newProvider();
+        }
+
+        private static final class ProviderFactory {
+            private final X509KeyManager keyManager;
+            private final String password;
+            private final Iterable<String> aliases;
+
+            ProviderFactory(X509KeyManager keyManager, String password, Iterable<String> aliases) {
+                this.keyManager = keyManager;
+                this.password = password;
+                this.aliases = aliases;
+            }
+
+            OpenSslKeyMaterialProvider newProvider() {
+                return new OpenSslPopulatedKeyMaterialProvider(keyManager,
+                        password, aliases.iterator());
+            }
+
+            /**
+             * {@link OpenSslKeyMaterialProvider} implementation that pre-compute the {@link OpenSslKeyMaterial} for
+             * all aliases.
+             */
+            private static final class OpenSslPopulatedKeyMaterialProvider extends OpenSslKeyMaterialProvider {
+                private final Map<String, Object> materialMap;
+
+                OpenSslPopulatedKeyMaterialProvider(
+                        X509KeyManager keyManager, String password, Iterator<String> aliases) {
```

Comment:
not sure I understand why the final arg is Iterator.     If this is a private class, and only has one callsite for the ctor, why not pass the iterable directly?   That would make the loop a for rather than a while, and not need to call `next` and `hasNext`

Reply:
@carl-mastrangelo it doesn't really matter to me... let me change it.